Decorators 2

albertjan opened the pull request:
Hi all here is the new decorators branch. But I still have a small issue. @meredydd  could you have a look at this:

```py
def decofunc(fn):
    return fn

@decofunc
def func():
    print "called"

func()
```

this fails because the `@decofunc` makes `func` into a suspension. And then it fails in `applyOrSuspend`.

I want to add the real tests from cpython for the property decorator.

closes #18 #426 #326

albertjan commented on src/abstract.js:
```diff
@@ -828,20 +828,24 @@ Sk.abstr.gattr = function (obj, nameJS, canSuspend) {
 
-    ret = Sk.misceval.chain(ret, function(ret) {
+    var eventualresult = Sk.misceval.chain(undefined, function () {
+        return obj;
```
@meredydd I've also changed this because the obj is a suspension aswell. But it may be that I should have done it differently, because it sortof feels like the same kind of issue as I mentioned in the PR.

albertjan replied:
I can revert this change now can't I?

meredydd commented on src/lib/image.js:
```diff
@@ -117,13 +117,37 @@ $builtinmodule = function (name) {
+
+            return new Sk.builtin.list(arr);
+        });
+
```
Why are these changes in here? Not sure what images have to do with decorators...

albertjan replied:
yeah they need to be gone 😄 I'll remove them

meredydd commented on src/misceval.js:
```diff
@@ -1189,6 +1189,14 @@ Sk.misceval.applyOrSuspend = function (func, kwdict, varargseq, kws, args) {
+        if (fcall !== undefined) {
+            fcall = Sk.misceval.callsimOrSuspend(func.__get__, func, func);
+            //args.unshift(func);
+            if (fcall.tp$call) {
```
This will fail if `func.__get__` returns a suspension. What are you trying to do here?

albertjan replied:
Well the problem we're having here is that `func` itself is a suspension, so it will throw an exception even before it gets there. this call was just a shot in the dark. Thats why I asked for your help 😄.

meredydd replied:
So, why is `func` a suspension? I'm guessing that it's because it's decorated, and the decorator suspended?

If this happens, we shouldn't continue and make the value of the function variable a suspension. What should happen is that the decorator should be allowed to complete before that variable gets assigned. Can you show me where the `compile.js` output for decorators is?

meredydd replied:
I'm either misreading Github's interface or this issue is still here - a call to `Sk.misceval.chain` should sort you out.

meredydd commented on src/object.js:
```diff
@@ -88,8 +102,28 @@ Sk.builtin.object.prototype.GenericSetAttr = function (name, value) {
+        // todo;
+        //if (f && descr.tp$descr_set) // is a data descriptor if it has a set
+        //return f.call(descr, this, this.ob$type);
+    }
```
What's `descr` short for here? I'm lost...

albertjan replied:
descriptor

albertjan replied:
they should probably have been a part of a different PR.

meredydd replied:
Can you separate them? It would certainly be easier to review, although I appreciate it's a pain in the ass.

meredydd commented on src/property.py:
```diff
@@ -0,0 +1,40 @@
+        return type(self)(self.fget, fset, self.fdel, self.__doc__)
+
+    def deleter(self, fdel):
+        return type(self)(self.fget, self.fset, fdel, self.__doc__)
```
Ah, so you are implementing property descriptors, then decorators to generate them? I'll keep hunting, so don't do anything yet, but it might help to break this into two PRs if that wouldn't be too much of a pain.

albertjan replied:
I agree.

meredydd replied:
Brill. Am happy to commit to reviewing both.

albertjan commented on src/compile.js:
```diff
@@ -1687,7 +1690,14 @@ Compiler.prototype.buildcodeobj = function (n, coname, decorator_list, args, cal
+        var res;
+        if (decos.length > 0)
+        {
+            res = this._gr("funcobj", "Sk.misceval.callsim(", scopename, ".$decorators[0], new Sk.builtins['function'](", scopename, ",$gbl", frees, "))"); // scopename, ".$decorators[0](new Sk.builtins['function'](", scopename, ",$gbl", frees, "))";
```
here the decorator is called.

meredydd commented on src/compile.js:
```diff
@@ -1687,7 +1690,14 @@ Compiler.prototype.buildcodeobj = function (n, coname, decorator_list, args, cal
+        var res;
+        if (decos.length > 0)
+        {
+            res = this._gr("funcobj", "Sk.misceval.callsimOrSuspend(", scopename, ".$decorators[0], new Sk.builtins['function'](", scopename, ",$gbl", frees, "))"); // scopename, ".$decorators[0](new Sk.builtins['function'](", scopename, ",$gbl", frees, "))";
```
Ah, here. We should do this in two stages:
1.  `out("$ret = Sk.misceval.callsimOrSuspend(", scopename, ".$decorators[0], new Sk.builtins['function'](", scopename, ",$gbl", frees, "))");`

2. `checkSuspension()`

3. return this._gr("funcobj", "$ret")

albertjan commented:
@meredydd that appears to have fixed the code example. 😄

albertjan commented:
and the tests @meredydd you're awesome :)

bnmnetp commented:
So, the current state is that the tests all pass in the non-closure compiled phase.  But after running the closure compiler, the unittests appear to get hung up in test_complex -- Travis kills them after 10 minutes.  They've been running for 15 on my macbook pro and still stuck in test_complex.

If I take test_complex out of the picture then all tests except for: `Tests failed in  TestGenericStringIO.test_input` pass.

Any ideas?

bnmnetp commented:
Just to update, test_complex does eventually finish after about 35 minutes.  But for some strange reason, test_input still fails.

waywaaard commented:
Testing complex takes ages with suspensions turned on for every line. Maybe I need to do some profiling to find the bottleneck.

bnmnetp commented:
Indeed, If I change `debug_mode` back to false then everything passes in a reasonable amount of time.

It seems to me that this would be a separate issue then.  Is it worth holding this one back over this or should we make a separate issue for the stringIO failing in the closure compiled phase with debug turned on?

rixner commented:
@bnmnetp, if you just check out master and do:

```
skulpt.py rundebug test/unit/test_StringIO.py
```

There is a failure.  So, I would argue that that problem should be investigated independently and has nothing to do with this pull request.  (I didn't look at the rest of this pull request, but just thought I'd point that out.)  Although, you should probably do this yourself and confirm that it's the same failure you are seeing...

albertjan commented:
Shall I turn debug_mode off then? And open a PR with that on, so we can add the changes to fix those issues to there?

bnmnetp commented:
@albertjan that would be my vote.  I would like to see this go in in the hope that we can get some implementation of property working next.

@rixner, I did that and saw the same failure.

albertjan commented:
@bnmnetp I've found the bug in input, and commented out the test in debug_mode but we really should run them. I'll try recompiling v8 the later versions are much faster.

bnmnetp commented:
@albertjan I agree they should be run, if they can be run fast enough to avoid having all our travis checks look like they have failed.

albertjan commented:
I agree, I'll look into building a newer version of v8 or using node to run skulpt (as node is just v8 with some stuff boulted on to it).

bnmnetp commented:
Yeah, I've often thought that it would be convenient if we could use node.

bnmnetp commented:
OK, @albertjan WOW!
So property is implemented in property.py and it just works!  😄 

We should probably remove the function definition with the unimplemented error from builtins.js ??

albertjan commented:
Ah good catch yeah we can remove that. Yes it "just works" any .py files placed into the src dir will be compiled during startup and assigned to `Sk.builtin.<name_of_file>`. I know we might want to change that behaviour in the future to assign to different things. But we can do that as and when needed.

meredydd commented on src/compile.js:
```diff
@@ -1687,7 +1690,16 @@ Compiler.prototype.buildcodeobj = function (n, coname, decorator_list, args, cal
+        var res;
+        if (decos.length > 0)
+        {
+            out("$ret = Sk.misceval.callsimOrSuspend(", scopename, ".$decorators[0], new Sk.builtins['function'](", scopename, ",$gbl", frees, "));");
```
Am I missing something obvious, or does this only support one decorator at a time?

albertjan commented:
Hey @meredydd I only see the comment about only the first decorator, I think you are correct there.

rixner commented:
This is pretty cool, so I hate to be the bad guy and complain about this yet again, but this one really concerns me.  I'm seeing up to a 20% slowdown from this pull request on programs that don't even use decorators.  In my opinion, that's a _lot_.

bnmnetp commented:
@rixner My guess is that it is the descriptor protocol more than the decorators themselves.  Descriptors could potentially impact accessing any instance variable.  I'm somewhat sympathetic to the performance argument, but I think this PR opens up so much more of Python to us!  I think this one enables lots more pure python modules to be added to the library.

How we make the tradeoffs of new functionality against performance is one that we need more discussion around.

meredydd commented:
I feel like 20% is bigger than should be associated with a change like this. Can you point a profiler at it, @rixner ?

albertjan commented:
It's compiling the property decorator at start up. Thats probably why its slower.

rixner commented:
I strongly suspect the problem is the new, expensive call to ```typeLookup``` at the start of ```GenericGetAttr```.  ```GenericGetAttr``` is a performance critical function and this is going to slow it down every time.  I'll profile it and confirm.

rixner commented:
Yes, ```Sk.builtin.type.typeLookup``` has become the number 1 function in the profile.

rixner commented:
@albertjan, when I measure time, I actually measure the body of the code and ignore compilation/startup overheads.  So, while that may have also gotten slower, I am personally not so concerned about that.

albertjan commented:
@rixner alright, it's obviously part of the GenericGetAttr. I'll look into speeding it up.

albertjan commented:
Why don't we move the typelookup to after the normal lookup so it will only do that when there's no normal property on the object. That would help with performance. probably.

albertjan commented:
@rixner do you have an example program that shows the performance hit? So I can run it locally?

rixner commented:
The gist I referenced [here](https://github.com/skulpt/skulpt/pull/597#issuecomment-231252466) in the comments for #597 is a program that shows a big slowdown for me.

albertjan commented:
Ok I've moved the typeLookup to after the normal field lookup. My understanding of descriptors in python aren't good enough to see if we break anything with this. The Performance is now the same as the master. Best run on my machine is 5.73 for both this branch and the master.

bnmnetp commented:
Here is a simple example:

```
class RevealAccess(object):
    """A data descriptor that sets and returns values
       normally and prints a message logging their access.
    """

    def __init__(self, initval=None, name='var'):
        self.val = initval
        self.name = name

    def __get__(self, obj):
        print('Retrieving', self.name)
        return self.val

    def __set__(self, obj, val):
        print('Updating', self.name)
        self.val = val


class MyClass(object):
    x = RevealAccess(10, 'var "x"')
    y = 5

m = MyClass()
print(m.x)
print(m.y)
m.x = 20
print(m.x)
```

my output is:
```
('Retrieving', 'var "x"')
10
5
('Updating', 'var "x"')
('Retrieving', 'var "x"')
20
```

rixner commented:
@albertjan, I also now see roughly the same performance as master.  Thanks!

albertjan commented:
I've merged it with master now and I see a performance increase of 0.3s on average due to the fast chain path. I'm still having one issue with the script @bnmnetp posted. My output differs:

```
('Retrieving', 'var "x"')
10
5
('Updating', 'var "x"')
20

('Retrieving', 'var "x"')
10
5
('Updating', 'var "x"')
('Retrieving', 'var "x"')
20
```

bnmnetp commented:
@albertjan so you are missing the second retrieving print statement even though it appears to have updated properly. Hmm.

albertjan commented:
I have to move the setter above the normal property accessor. I'm still seeing the same performance.

bnmnetp commented:
@albertjan This is so great.  

Here's another little test program.  It was working earlier today but is broken on the latest.

```
class P:

    def __init__(self,x):
        self.x = x

    @property
    def x(self):
        return self.__x

    @x.setter
    def x(self, x):
        if x < 0:
            self.__x = 0
        elif x > 1000:
            self.__x = 1000
        else:
            self.__x = x

t = P(9)
print(t.x)
t.x = 100000
print(t.x)
```

I get ``UNCAUGHT EXCEPTION: ExternalError: TypeError: Cannot read property '0' of undefined on line 6``

Instead of
```
9
1000
```

bnmnetp commented:
Ah, I see that ``test/unit/test_decorators.py`` went missing or the unit tests would have caught that.

albertjan commented:
Right ... :) I'll see if I can find that file, and if I can fix the bug.

bnmnetp commented:
Hi @albertjan,

Any progress on this one?  Are we stuck with a tradeoff of a slow but working implementation, or fast but non-working?

albertjan commented:
@bnmnetp yeah pretty much, I'll add the tests back in, and return it to working state, see if we can optimise it in a different way.

albertjan commented:
@bnmnetp I accidentally removed some code from the compiler causing this problem we saw. Do you know where in the cpython test suite the decorators and descriptors are tested?

bnmnetp commented:
@albertjan At one point this PR contained the file test/unit/test_decorators.py

You should be able to recover it from the history.

albertjan commented:
yeah I have it 😄 thanks

albertjan commented:
I'm trying to get the `test_decorators.py` to work but it seems expertly crafted to show all skulpt's creavats. https://github.com/python/cpython/blob/2.7/Lib/test/test_decorators.py

bnmnetp commented:
If I'm reading my log correctly all of those tests were passing from a branch called trinketapp-decorators-2 at this point:  28e31ea039dff0e4c059d1221ceab9b81263f6ca

albertjan commented:
@bnmnetp they can't have been they use double decorators and the compile function in almost all of the tests. they use `func_name`. all of which aren't supported.  I've added your version of the tests back in. I've changed them a little to make sure they also at least compile on cpython.

albertjan commented:
Good find @bnmnetp I thought I'd lost them in the rebase. I don't know from which branch in the cpython repo I got them. I've added them next to the other tests in the repo, to be extra sure it works :smile:. If this all builds and @rixner is happy with the perf, i'd say it's ready to be merged :shipit:.

We should note that we currently support only one decorator.

rixner commented:
What is internalpython.js and the other seemingly associated stuff (like ```importBuiltinWithBody```) in here?  And what does that have to do with decorators?

albertjan commented:
internalpython.js is the file that contains the parts of the stdlib that are written in python, for now only property.py. importBuiltinWithBody compiles these parts.

rixner commented:
Perhaps I am missing something really obvious, but why is this necessary?  If you put property.py in the lib directory, it will be importable.  This already works, why do we need a second mechanism to do the same thing?

albertjan commented:
`property` is a builtin like `map` and `filter` they don't need to imported. Thats why we need to do it when we initialise skulpt.

albertjan commented:
they aren't imported into `__main__` either

rixner commented:
Ah, sorry!  I was missing something really obvious. :smile:

bnmnetp commented:
I just merged this against the current master, and ran all the tests. While I did not measure the performance impact, it was able to take into account the gains from #604 and so seemed quite speedy.

albertjan commented:
😄 thats good, we do take a small hit during initialisation, that we may want to move forward to compile time with #609.

bnmnetp commented:
Are we all happy (enough) with this one to merge it?

bnmnetp commented:
A bit more empirical, I just timed ``./skulpt.py rununits``

This PR
```
Passed:   292 Failed     0
       29.49 real        29.01 user         1.50 sys
```

master
```
Passed:   289 Failed     0
       25.58 real        25.14 user         1.44 sys
```

I think we should merge this to get the feature into the language, but this gives us more to work on in terms of looking for optimizations.

albertjan commented:
there are some optimisations on the main branch that aren't on this one so the numbers may be a bit skewed by that.

bnmnetp commented:
OK, unless I hear otherwise, I'm going to merge this later today.

I think we've identified `GenericGetAttr` as an important starting point for looking at Optimizations, but the descriptor protocol is an integral part of that and it would make no sense to seriously look at Optimizing that without this feature in place.

rixner commented:
I still think this is a large slowdown.

I concede, however, that I care more about not losing any further performance than I do about this particular feature.  If others feel differently, then so be it.

albertjan commented:
@rixner let me do a merge with master so we can objectivly compare the perf difference.

albertjan commented:
this branch:

rununits:
```sh
./m rununits  25.04s user 1.32s system 106% cpu 24.841 total
./m rununits  25.10s user 1.28s system 106% cpu 24.795 total
./m rununits  25.46s user 1.27s system 106% cpu 25.140 total
```
field.py
```sh
time elapsed: 5.60977315903
./m run field.py  6.40s user 0.13s system 100% cpu 6.478 total

time elapsed: 5.49832081795
./m run field.py  6.30s user 0.12s system 100% cpu 6.380 total

time elapsed: 5.55619883537
./m run field.py  6.37s user 0.13s system 100% cpu 6.454 total
```

master:

```sh
./m rununits  24.33s user 1.21s system 106% cpu 24.019 total
./m rununits  23.75s user 1.20s system 106% cpu 23.472 total
./m rununits  24.02s user 1.19s system 106% cpu 23.705 total
```

field.py:
```sh
time elapsed: 5.57643413544
./m run field.py  6.32s user 0.12s system 100% cpu 6.406 total

time elapsed: 5.67489504814
./m run field.py  6.46s user 0.12s system 100% cpu 6.533 total

time elapsed: 5.81251382828
./m run field.py  6.57s user 0.12s system 100% cpu 6.647 total
```

bnmnetp commented:
Wow, thanks @albertjan !  I'll bet even @rixner can live with that 😄

LeszekSwirski commented on src/import.js:
```diff
@@ -539,6 +547,21 @@ Sk.importMainWithBody = function (name, dumpJS, body, canSuspend) {
+/**
+ * **Run Python Code in Skulpt**
+ *
+ * When you want to hand Skulpt a string corresponding to a Python program this is the function.
```
I thought this was the job of importMainWithBody?

rixner replied:
@albertjan, since both @LeszekSwirski and I were confused by this, I think it would help us and all future developers if this comment were improved to more clearly explain the purpose of this function and its relation to the other import functions...  The current comment is misleading, in my opinion.

rixner commented:
Yes, I can live with that.  Clearly, the overhead of this pull request is now quite low!
